carla/recourse_methods/catalog/cem/model.py

```diff
@@ -179,28 +123,101 @@ def __init__(self, sess, catalog_model: MLModel, hyperparams):
             0,
             power=0.5,
         )
-        optimizer = tf.train.GradientDescentOptimizer(learning_rate)
+
+        self.train = self.__optimization(Loss_ToOptimize_s, learning_rate)
         start_vars = set(x.name for x in tf.global_variables())
-        self.train = optimizer.minimize(
-            Loss_ToOptimize,
-            var_list=[self.adv_img_s],
-            global_step=self.global_step,
-        )
-        end_vars = tf.global_variables()
-        new_vars = [x for x in end_vars if x.name not in start_vars]
+        new_vars = [x for x in tf.global_variables() if x.name not in start_vars]
 
         # these are the variables to initialize when we run
-        self.setup = []
-        self.setup.append(self.orig_img.assign(self.assign_orig_img))
-        self.setup.append(self.target_lab.assign(self.assign_target_lab))
-        self.setup.append(self.const.assign(self.assign_const))
-        self.setup.append(self.adv_img.assign(self.assign_adv_img))
-        self.setup.append(self.adv_img_s.assign(self.assign_adv_img_s))
+        self.setup = self.__set_setup()
 
         self.init = tf.variables_initializer(
             var_list=[self.global_step] + [self.adv_img_s] + [self.adv_img] + new_vars
         )
 
+    def __optimization(self, Loss_ToOptimize_s, learning_rate):
```

Comment:
Leading double underscore should only be used for _name mangleling_ cases. If you want to indicate that this function is "private" you should use the single underscore
- https://www.python.org/dev/peps/pep-0008/#id47
- https://docs.python.org/3/tutorial/classes.html#private-variables

Reply:
Right, will change this 👍